File: goldencheck/engine/db_scanner.py
"""Database scanner — scan tables directly from Postgres, Snowflake, BigQuery, or any SQLAlchemy URL."""
from __future__ import annotations

import logging
import tempfile
from pathlib import Path

import polars as pl

from goldencheck.engine.scanner import scan_file
from goldencheck.engine.confidence import apply_confidence_downgrade
from goldencheck.models.finding import Finding
from goldencheck.models.profile import DatasetProfile

logger = logging.getLogger(__name__)

__all__ = ["scan_database"]


def scan_database(
    connection_string: str,
    table: str | None = None,
    query: str | None = None,
    sample_size: int = 100_000,
    domain: str | None = None,
) -> tuple[list[Finding], DatasetProfile]:
    """Scan a database table or query result.

    Args:
        connection_string: Database URL (postgres://, snowflake://, bigquery://, etc.)
        table: Table name to scan (SELECT * FROM table LIMIT sample_size)
        query: Custom SQL query (overrides table)
        sample_size: Max rows to fetch
        domain: Domain pack name

    Returns:
        (findings, profile) tuple
    """
    if not table and not query:
        raise ValueError("Either 'table' or 'query' must be provided.")

    sql = query or f"SELECT * FROM {table} LIMIT {sample_size}"

    # Try Polars native connectors first
    df = _read_sql(connection_string, sql)

    # Write to temp CSV and scan (reuses the full profiler pipeline)
    with tempfile.NamedTemporaryFile(suffix=".csv", delete=False, mode="w") as tmp:
        df.write_csv(tmp.name)
        tmp_path = Path(tmp.name)

    try:
        findings, profile = scan_file(tmp_path, sample_size=sample_size, domain=domain)
        findings = apply_confidence_downgrade(findings, llm_boost=False)
    finally:
        tmp_path.unlink(missing_ok=True)

    # Fix the file_path in profile to show the actual source
    source = table or "custom query"
    profile = DatasetProfile(
        file_path=f"{_mask_password(connection_string)}:{source}",
        row_count=profile.row_count,
        column_count=profile.column_count,
        columns=profile.columns,
    )

    return findings, profile


def _read_sql(connection_string: str, sql: str) -> pl.DataFrame:
    """Read SQL query into a Polars DataFrame."""
    # Try connectorx (fastest, supports most databases)
    try:
        import connectorx as cx
        return cx.read_sql(connection_string, sql, return_type="polars")
    except ImportError:
        pass

    # Try SQLAlchemy + pandas fallback
    try:
        import sqlalchemy
        import pandas as pd
        engine = sqlalchemy.create_engine(connection_string)
        with engine.connect() as conn:
            pdf = pd.read_sql(sql, conn)
        return pl.from_pandas(pdf)
    except ImportError:
        pass

    raise ImportError(
        "Database scanning requires 'connectorx' or 'sqlalchemy+pandas'. "
        "Install with: pip install connectorx  (or)  pip install sqlalchemy pandas"
    )


def _mask_password(url: str) -> str:
    """Mask password in connection string for display."""
    import re
    return re.sub(r'://([^:]+):([^@]+)@', r'://\1:***@', url)

File: packages/goldencheck-js/src/node/db-scanner.ts
/**
 * Database scanner — scan tables directly from Postgres, MySQL, or SQLite.
 * Port of goldencheck/engine/db_scanner.py. Node-only.
 *
 * Database drivers are optional peer dependencies:
 * - Postgres: `pg`
 * - MySQL: `mysql2`
 * - SQLite: `better-sqlite3`
 */

import { TabularData, type Row } from "../core/data.js";
import { scanData } from "../core/engine/scanner.js";
import { applyConfidenceDowngrade } from "../core/engine/confidence.js";
import type { ScanResult } from "../core/types.js";

export async function scanDatabase(
  connectionString: string,
  options?: {
    table?: string;
    query?: string;
    sampleSize?: number;
    domain?: string;
  },
): Promise<ScanResult> {
  const table = options?.table;
  const query = options?.query;
  const sampleSize = options?.sampleSize ?? 100_000;
  const domain = options?.domain;

  if (!table && !query) {
    throw new Error("Either 'table' or 'query' must be provided.");
  }

  const sql = query ?? `SELECT * FROM ${table} LIMIT ${sampleSize}`;

  // Determine driver from connection string and fetch rows
  const rows = await executeQuery(connectionString, sql);

  if (rows.length === 0) {
    const data = new TabularData([]);
    const result = scanData(data, { sampleSize, domain: domain ?? undefined });
    return {
      findings: applyConfidenceDowngrade(result.findings, false),
      profile: {
        ...result.profile,
        filePath: `${maskPassword(connectionString)}:${table ?? "custom query"}`,
      },
    };
  }

  const data = new TabularData(rows);
  const result = scanData(data, { sampleSize, domain: domain ?? undefined });
  const findings = applyConfidenceDowngrade(result.findings, false);

  // Replace file path with source info (masked connection string)
  const source = table ?? "custom query";
  const profile = {
    ...result.profile,
    filePath: `${maskPassword(connectionString)}:${source}`,
  };

  return { findings, profile };
}

// --- Driver detection & query execution ---

type DriverType = "postgres" | "mysql" | "sqlite";

function detectDriver(connectionString: string): DriverType {
  if (connectionString.startsWith("postgres://") || connectionString.startsWith("postgresql://")) {
    return "postgres";
  }
  if (connectionString.startsWith("mysql://")) {
    return "mysql";
  }
  if (connectionString.startsWith("sqlite://") || connectionString.startsWith("sqlite3://")) {
    return "sqlite";
  }
  throw new Error(
    `Unsupported database URL scheme. Expected postgres://, mysql://, or sqlite://. Got: ${connectionString.split("://")[0]}://`,
  );
}

async function executeQuery(connectionString: string, sql: string): Promise<Row[]> {
  const driver = detectDriver(connectionString);

  switch (driver) {
    case "postgres":
      return executePostgres(connectionString, sql);
    case "mysql":
      return executeMysql(connectionString, sql);
    case "sqlite":
      return executeSqlite(connectionString, sql);
  }
}

async function executePostgres(connectionString: string, sql: string): Promise<Row[]> {
  // eslint-disable-next-line @typescript-eslint/no-require-imports
  let pg: any;
  try {
    pg = require("pg");
  } catch {
    throw new Error(
      "Postgres scanning requires the 'pg' package. Install with: npm install pg",
    );
  }

  const Client = pg.default?.Client ?? pg.Client;
  const client = new Client({ connectionString });

  try {
    await client.connect();
    const result = await client.query(sql);
    return result.rows as Row[];
  } finally {
    await client.end();
  }
}

async function executeMysql(connectionString: string, sql: string): Promise<Row[]> {
  // eslint-disable-next-line @typescript-eslint/no-require-imports
  let mysql2: any;
  try {
    mysql2 = require("mysql2/promise");
  } catch {
    throw new Error(
      "MySQL scanning requires the 'mysql2' package. Install with: npm install mysql2",
    );
  }

  const connection = await mysql2.createConnection(connectionString);

  try {
    const [rows] = await connection.execute(sql);
    return rows as Row[];
  } finally {
    await connection.end();
  }
}

function executeSqlite(connectionString: string, sql: string): Promise<Row[]> {
  const dbPath = connectionString.replace(/^sqlite3?:\/\//, "");

  // eslint-disable-next-line @typescript-eslint/no-require-imports
  let Database: any;
  try {
    Database = require("better-sqlite3");
  } catch {
    throw new Error(
      "SQLite scanning requires the 'better-sqlite3' package. Install with: npm install better-sqlite3",
    );
  }

  const db = Database(dbPath);
  try {
    const rows = db.prepare(sql).all() as Row[];
    return Promise.resolve(rows);
  } finally {
    db.close();
  }
}

// --- Helpers ---

/** Mask password in connection string for display. */
function maskPassword(url: string): string {
  return url.replace(/:\/\/([^:]+):([^@]+)@/, "://$1:***@");
}
